### ISSUE TITLE
Blazor: @onclick callback not being invoked, when element is within foreach and with @key

### ISSUE THREAD

**Issue by Liero** [dotnet/aspnetcore#29985]
### Describe the bug
I found this strange behavior when I have multiple DOM events invoked (i.e `blur` and `click`):
When the first event (blur) modifies a collection, that the latter (click) won't be invoked on elements within `@foreach` loop:

see repro: https://github.com/Liero/aspnetcore-blazor_onclick_after_onblur

```razor
<input @onblur="OnBlur" @oninput="OnInput" />
<div style="height: 20em; overflow: auto">
    @foreach(var item in Items) 
    {
        <!-- BUG? OnItemClick not invoked, when user click item while focus is in the input  -->
        <div @onclick="_ => OnItemClick(item)" @key="item">@item</div>
    }
</div>

@code {
    List<string> Items = GenerateItems().ToList();

    void OnBlur() {
        Items = GenerateItems().ToList();
    }

    void OnInput(ChangeEventArgs args) {
        Items = GenerateItems()
            .Where(i => i.Contains((string)args.Value, StringComparison.CurrentCultureIgnoreCase))
            .ToList();
    }

    void OnItemClick(string item) {
        //not invoked, when OnBlur called first
        SelectedItem = item;
    }

    static IEnumerable<string> GenerateItems() => Enumerable.Range(0, 40)
        .Select(i => $"{i} " + new string("ABCDEFGHIJKL".Skip(i % 10).Take(4).ToArray()));
}
```

If this is by design, then what is a workaround?

### Special Comment
Notice, that if I remove `@key="item"` from the div, the OnItemClick eventhandler will be invoked, but with wrong parameter. 

### To Reproduce

1. Run code above, or use https://blazorrepl.com/repl/GbkQEMbG52kiwSuD01
2. Type 'C' into the textbox
3. Click the 4th item, while textbox is focused.

![image](https://i.imgur.com/EnK7aTJ.gif)

### Further technical details
- ASP.NET Core version 5.0
- Blazor Server Hosting Mode.
- `dotnet --info`
```
.NET SDK (reflecting any global.json):
 Version:   5.0.102
 Commit:    71365b4d42

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.102\

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  2.1.801 [C:\Program Files\dotnet\sdk]
  5.0.100-rc.2.20479.15 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-rc.2.20475.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-rc.2.20475.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.2.20475.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by javiercn** [comment]
@Liero thanks for contacting us.

You can use the following instructions to determine if the event is indeed not being fired.

Add this to your _Host.cshtml/Index.html
```razor
    <script src="_framework/blazor.server.js" autostart="false"></script>
    <script>
        (function start() {
            Blazor.start({
                logLevel: 1, // LogLevel.Debug
                // Omit this part for WASM
                configureSignalR: builder => builder.configureLogging("debug") // LogLevel.Debug
            });
        })()
    </script>
```

Replace the logging section with this in your appSettings.json
```json
{
  "Logging": {
    "LogLevel": {
      "Default": "Debug",
      "System": "Debug",
      "Microsoft": "Debug"
    }
  }
}

Could you check that and let us know if that's the case?

I would expect to see logs on the console indicating the event happened or a new render showing after clicking the button.
```

**Comment by Liero** [comment]
@javiercn:
there is no click event being dispatched, indeed. Following logs are emitted, when I click the div, while textbox is focused:
```
Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher: Debug: Received hub invocation: InvocationMessage { InvocationId: "", Target: "DispatchBrowserEvent", Arguments: [ {"browserRendererId":0,"eventHandlerId":1,"eventArgsType":"focus","eventFieldInfo":{"componentId":0,"fieldValue":"a"}}, {"type":"blur"} ], StreamIds: [  ] }.
Microsoft.AspNetCore.Components.RenderTree.Renderer: Debug: Handling event 1 of type 'FocusEventArgs'
Microsoft.AspNetCore.Components.RenderTree.Renderer: Debug: Rendering component 0 of type BlazorApp1.App
Microsoft.AspNetCore.Components.Server.Circuits.RemoteRenderer: Debug: Sending render batch 11 of size 3431 bytes to client 0Boq-DD25y0Es4Yue1nG9g.
Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher: Debug: Received hub invocation: InvocationMessage { InvocationId: "", Target: "OnRenderCompleted", Arguments: [ 11,  ], StreamIds: [  ] }.
Microsoft.AspNetCore.Components.Server.ComponentHub: Debug: Received confirmation for batch 11
Microsoft.AspNetCore.Components.Server.Circuits.RemoteRenderer: Debug: Completing batch 11 without error in 16.8879ms.
```
Please, notice, that the div is removed from the DOM in onblur event. If it is not removed, then the click event is there.
Also notice, that when I remove `@key="item"` from the div, the click event is dispatched, but with wrong parameter.
I've also attached screen recording in the original post

**Comment by javiercn** [comment]
@Liero thanks for the additional details.

We'll take a look at this in more detail and and update this issue accordingly.

**Comment by mkArtakMSFT** [comment]
Looks like this is now fixed in the 6.0 release.
